File: Algorithms/p115_Distinct_Subsequences/p115_Distinct_Subsequences.py
class Solution(object):
    def numDistinct(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: int
        """
        m = len(s)
        n = len(t)
        num = [[0 for j in range(n+1)] for i in range(m+1)]
        num[0][0] = 1
        for i in range(1,m+1):
            num[i][0] = 1
            for j in range(1,min(i,n)+1):
                num[i][j] = num[i-1][j]
                if s[i-1]==t[j-1]:
                    num[i][j] += num[i-1][j-1]
        return num[m][n]

File: Algorithms/p115_Distinct_Subsequences/p115_Distinct_Subsequences.cpp
class Solution {
public:
    int numDistinct(string s, string t) {
        int m = s.size();
        int n = t.size();
        vector<vector<int>> num(m+1,vector<int>(n+1,0));
        num[0][0] = 1;
        for(int i=1;i<=m;++i){
            num[i][0] = 1;
            for(int j=1;j<=i&&j<=n;++j){
                num[i][j] = num[i-1][j];
                if(s[i-1]==t[j-1]) num[i][j] += num[i-1][j-1];
            }
        }
        return num[m][n];
    }
};
